### PR TITLE
Snowflake cdc

### PR DESCRIPTION
**Pull request by Ismail-Naziz**
## What
CDC Implementation for all this cases:
cdc full refresh
cdc incremental
cdc pushdown filters (both incremental and full refesh)

## How
I created new classes for cdc that behave like streams with a difference in the statement

## Review guide
look at streams.py and source.py

## User Impact
User can now chose how to increment data

## Can this PR be safely reverted and rolled back?
YES 💚

# Required
You must enable the change tracking to use CDC
with this sql statement 
-- Enable change tracking on the table.
 ALTER TABLE t1 SET CHANGE_TRACKING = TRUE;
You can have more details about changes here: https://docs.snowflake.com/en/sql-reference/constructs/changes

### CODE REVIEW

File: airbyte-integrations/connectors/source-snowflake/source_snowflake/streams.py [hunk 1]
```diff
@@ -633,3 +634,129 @@ def statement(self):
...
+        ordered_mapping_names_types[self.DEFAULT_CURSOR_FIELD] = 'TIMESTAMP_TZ'
+        current_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+
+        for record in response_json.get("data", []):
+            record.append(current_time)
+            yield {column_name: format_field(column_value, ordered_mapping_names_types[column_name])
+                   for column_name, column_value in zip(ordered_mapping_names_types.keys(), record)}
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        # TODO: clean by using parent method
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-snowflake/source_snowflake/streams.py:703]: noting the TODO here

**Review by maxi297 — COMMENTED** [review]
I can't tests as I'm getting `422 Client Error: Unprocessable Entity for url` with statement `"SELECT * FROM "ALEX_TEST_DB"."POSTHOG"."ADACCOUNTS_STATS_DAILY" CHANGES(INFORMATION => DEFAULT) AT(TIMESTAMP => TO_TIMESTAMP('2024-05-20 11:12:23'))"`

File: airbyte-integrations/connectors/source-snowflake/source_snowflake/source.py [hunk 2]
```diff
@@ -82,14 +83,41 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
...
+
+        standard_streams = self._get_standard_streams(stream_class, config, url_base, table_catalog_stream, authenticator)
+        push_down_filters_streams = self._get_push_down_filters_streams(push_down_filters_class,
+                                                                        config, url_base, authenticator, standard_streams)
+        streams = list(standard_streams.values()) + push_down_filters_streams
+
+        return streams
+
+    @classmethod
+    def _get_standard_streams(cls, stream_class, config, url_base, table_catalog_stream, authenticator):
```
[the remark this answers is not in the document]
Comment (Ismail-Naziz) [reply airbyte-integrations/connectors/source-snowflake/source_snowflake/source.py:108]: Can make sure you enabled changes on the table ? It had worked for me with the exact same query (and changing the table of course)

File: airbyte-integrations/connectors/source-snowflake/source_snowflake/streams.py [hunk 3]
```diff
@@ -456,12 +457,12 @@ def _get_updated_state(self, latest_record: Mapping[str, Any]) -> Mapping[str, A
...
             self._cursor_value = max(latest_record_state, current_state_value) if current_state_value is not None else latest_record_state
             self.state = {self.cursor_field: self._cursor_value}
-            return {self.cursor_field: self._cursor_value}
-        self._cursor_value = latest_record_state
-        self.state = {self.cursor_field: self._cursor_value}
+        else:
+            self._cursor_value = latest_record_state
+            self.state = {self.cursor_field: self._cursor_value}
 
         self.logger.info(f"current state of {self.name} is {self.state}")
```
[the remark this answers is not in the document]
Comment (Ismail-Naziz) [reply airbyte-integrations/connectors/source-snowflake/source_snowflake/streams.py:464]: indeed it's noisy, I remove this log

**Pull request closed without merging**